File: 438-find-all-anagrams-in-a-string/find-all-anagrams-in-a-string.py
class Solution:
    def findAnagrams(self, s: str, p: str) -> List[int]:
        if len(p) > len(s):
            return []
        cnt_rep = Counter(p)
        k = len(p)
        sub  = Counter()
        for i in range(k):
            sub[s[i]] +=1
        ans = []
        if sub == cnt_rep:
            ans.append(0)
        for i in range(k, len(s)):
            sub[s[i]] +=1
            if sub[s[i-k]] > 1:
                sub[s[i-k]]  -=1
            else:
                del sub[s[i-k]] 
            if sub == cnt_rep:
                ans.append(i-k+1)
        return ans

File: 438-find-all-anagrams-in-a-string/find-all-anagrams-in-a-string.cpp

class Solution {
public:
    vector<int> findAnagrams(string s, string p) {
        vector<int> result;
        if (s.size() < p.size()){
            return result;
        }

        vector<int> p_count(26, 0), s_count(26, 0);
        for (char c : p) p_count[c - 'a']++;
        
        for (int i = 0; i < s.size(); i++) {
            s_count[s[i] - 'a']++;
            if (i >= p.size()) s_count[s[i - p.size()] - 'a']--;
            if (s_count == p_count) result.push_back(i - p.size() + 1);
        }
        
        return result;
    }
};